### PR TITLE
Rework multi criteria and add assets categorization

### PR DESCRIPTION
**Pull request by gillesdubois**
## Status
**READY**

## Mentions
Thx to @TiVince for the design idea

## Description
Complete refactor of the multi criteria search that was completely obsolete

This rework might look a bit more complex than the initial multi search but is way more modular than the older search.

Major improvements : 
- Code refactor 
- Display / processing split
- Manage all table (plugins included)
- Add select2 librarie for future multi search update

TODO : 
* Add accountinfo 
* Add GET interface 
* Manage OR / AND better

## Related Issues
- #413
- #378
- #377 
- #375
- #369
- #367
- #365 
- #358  
- #355 

## Todos
- [x] Tests

## Deploy Notes
Notes regarding deployment the contained body of work.  These should note any
db migrations, logical changes, etc.

1. One DB Migration to add assets categorization (hardware modificaition + new table)

## Impacted Areas in Application
List general components of the application that this PR will affect:

* Multi search
* New sub menu in "manage"

### CODE REVIEW

File: require/search/DatabaseSearch.php [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+        foreach ($idList as $index => $fields) {
+            $idArray[] = $fields['hardwareID'];
+        }
+        return $idArray;
+    }
+
+    /**
+     * Undocumented function
+     *
+     * @param [type] $type
```
Comment (MyTheValentinus) [inline require/search/DatabaseSearch.php]: Array ? in param ?

Comment (gillesdubois) [reply require/search/DatabaseSearch.php]: PHP Docs added

**Comment by MyTheValentinus** [comment]
PR is ok for me with last modifications !

**Comment by TiVince** [comment]
looks very promising !!
do you still support PHP 5.4 ?
this kind of syntax for empty() will not work before PHP5.5
https://github.com/OCSInventory-NG/OCSInventory-ocsreports/blob/3d265bf749648794deaab29b1af049ed22ad8d15/require/search/TranslationSearch.php#L217

see
https://stackoverflow.com/questions/1075534/cant-use-method-return-value-in-write-context

I found 3 functions with this problem in **TranslationSearch.php**
I had to modify this way to avoid  the error message
```
 public function getTranslationFor($name){
        global $l;
        $name = strtolower($name);
        $translation =$l->g($this->translationArray[$name]);
        //if(empty($l->g($this->translationArray[$name]))){
        if (empty($translation)){
            return $name;
        }
        return $l->g($this->translationArray[$name]);
    }

```


```
public function getTranslationForListField($string){
        global $l;

        $values = explode(".", $string);
        $translation =  $l->g($this->translationArray[$values[0]]);
       // if(!empty($l->g($this->translationArray[$values[0]]))){
        if ( ! empty($translation) ){
            $table = $translation;
        }else{
            $table = $values[0];
        }
```
```
 public function getTranslationForListField($string){
        global $l;

        $values = explode(".", $string);
        $translation =  $l->g($this->translationArray[$values[0]]);
       // if(!empty($l->g($this->translationArray[$values[0]]))){
        if ( ! empty($translation) ){
            $table = $translation;
        }else{
            $table = $values[0];
        }

        $name = strtolower($values[1]);
        $translation= $l->g($this->translationArray[$name]);
        if(!empty($translation)){
            $name = $l->g($this->translationArray[$name]);
        }

        return $table." : ".$name;

    }

```

**Comment by TiVince** [comment]
there is an error syntax  in  7013 and 7014 SQL files which prevent the scripts to run 

exemple 

```
> cat 7013.sql |mysql  -uroot -p ocsweb
> Enter password: 
> ERROR 1064 (42000) at line 14: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '--Add CATEGORY column to software table' at line 1
```

https://github.com/OCSInventory-NG/OCSInventory-ocsreports/blob/3d265bf749648794deaab29b1af049ed22ad8d15/files/update/7013.sql#L14
  a whitespace is missing
-- Add CATEGORY column to software table'

https://github.com/OCSInventory-NG/OCSInventory-ocsreports/blob/3d265bf749648794deaab29b1af049ed22ad8d15/files/update/7014.sql#L11

 -- Add CATEGORY column to hardware table

**Comment by gillesdubois** [comment]
Hi,

We support PHP 5.4 in the last 2.4.1 but i think the 2.5 will require PHP 5.5 minimum.
PHP 5.5 integrate a lot of feature we plan to use in the near future. 
Of course if you find it's a bad idea we can debate about it.

What do you think ?

The sql file bug has been fixed in an other pull request by @charleneauger 

@TiVince : You are contributing a lot to our software. I spoke to the OCS team and we would like to integrate you in our contributing team. If the offer interest you, you can send me a mail to gilles.dubois@ocsinventory-ng.org. 

Regards,
Gilles Dubois.

**Pull request merged**